tests/common/devices.py

```diff
@@ -1301,6 +1301,122 @@ def execute (self, cmd) :
         if (self.os == 'ixia') :
             eval(cmd)
 
+import copy
+
+class SonicAsic(object):
+
+    def __init__(self, sonichost, asic_index):
+        self.sonichost = sonichost
+        self.asic_index = asic_index
+
+    # Wrapper for ASIC/namespace aware modules
+    def bgp_facts(self, *module_args, **complex_args):
+        if self.sonichost.facts['num_asic'] != 1:
+            complex_args['instance_id'] = self.asic_index
+        return self.sonichost.bgp_facts(*module_args, **complex_args)
+
+
+class MultiAsicSonicHost(object):
+
+    def __init__(self, ansible_adhoc, hostname):
+        self.sonichost = SonicHost(ansible_adhoc, hostname)
+        self.asics = [SonicAsic(self.sonichost, asic_index) for asic_index in range(self.sonichost.facts["num_asic"])]
+
+    def _run_on_asics(self, *module_args, **complex_args):
+        if "asic_index" not in complex_args:
+            # Default ASIC/namespace
+            return getattr(self.sonichost, self.multi_asic_attr)(*module_args, **complex_args)
+        else:
+            asic_complex_args = copy.deepcopy(complex_args)
+            asic_index = asic_complex_args.pop("asic_index")
+            if type(asic_index) == int:
+                # Specific ASIC/namespace
+                if self.sonichost.facts['num_asic'] == 1:
+                    if asic_index != 0:
+                        raise ValueError("Trying to run module '{}' against asic_index '{}' on a single asic dut '{}'".format(self.attr, asic_index, self.sonichost.hostname))
+                return getattr(self.asics[asic_index], self.multi_asic_attr)(*module_args, **asic_complex_args)
+            elif type(asic_index) == str and asic_index.lower() == "all":
+                # All ASICs/namespace
+                if self.sonichost.facts['num_asic'] == 1:
+                    return [getattr(asic, self.multi_asic_attr)(*module_args, **asic_complex_args) for asic in self.asics]
```

Comment:
I have the config_facts ansible module also prototyped, but would commit that as part of a seperate PR. The method in SonicAsic for config_facts took care of adding 'host' module param if it is not specified. This was to address the case where we make config_facts call on a multi-dut DutHosts, where the value of 'host' was hard to define for each node.

```
    def config_facts(self, *module_args, **complex_args):
        if 'host' not in complex_args:
            complex_args['host'] = self.sonichost.hostname
        if self.sonichost.facts['num_asic'] != 1:
            complex_args['namespace'] = 'asic{}'.format(self.asic_index)
        return self.sonichost.config_facts(*module_args, **complex_args)
```

The issue with the above is when we make the call:

```
# Make call w/out host - the config_facts method in SonicAsic will take care of adding host ansible module.
duthosts.frontend_nodes.config_facts(asic_index='all')
```
With your method, it would fail as host is not specified.  So, looping through the asics fixes the above. 

In my last commit, i have removed the if check for 'num_asic' == 1, as the call when 'num_asic' is greater than 1.

Reply:
You are right. The code in your last commit looks like the right solution.